sktime/dists_kernels/_base.py

```diff
@@ -0,0 +1,251 @@
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+
+        Writes to self
+        --------------
+        symmetric: bool = True if X2 was not passed, False if X2 was passed
+            for use to make internal calculations efficient, e.g., in _transform
+        """
+        # no input checks or input logic here, these are done in transform
+        # this just defines __call__ as an alias for transform
+        return self.transform(X=X, X2=X2)
+
+    def transform(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2 (equal to X if not passed)
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+
+        Writes to self
+        --------------
+        symmetric: bool = True if X2 was not passed, False if X2 was passed
+            for use to make internal calculations efficient, e.g., in _transform
+        """
+
+        X = check_series(X)
+
+        if X2 is None:
+            X2 = X
+            self.symmetric = True
+        else:
+            X2 = check_series(X2)
+            self.symmetric = False
+
+        return self._transform(X=X, X2=X2)
+
+    def _transform(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2 (equal to X if not passed)
+
+            core logic
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+        """
+        raise NotImplementedError
+
+    def fit(self, X=None, X2=None):
+        # no fitting logic, but in case fit is called or expected
+        pass
+
+
+def _pairwise_panel_x_check(X):
+    """
+    Method used to check the input and convert
+    numpy 3d or numpy list of df to list of dfs
+
+    Parameters
+    ----------
+    X: List of dfs, Numpy of dfs, 3d numpy
+        The value to be checked
+
+    Returns
+    -------
+    X:
+        Checked and converted to List of pd.Dataframe if one
+        of the other formats
+
+    """
+
+    def arr_check(arr):
```

Comment:
Minor comment:
seems odd to have function defined within function. any particular reason not to have def arr_check as a separate helper function?

Reply:
So this is more my style of programming than anything so happy to get feedback on what people think about it. The reason it's defined within the scope is because I feel as a stand alone function it's not very useful and only when combined with the rest of the checks going on here does it have value. The reason it is a function then rather than just lines of code is because it is called when certain conditions and the conditions can't ever occur together and so rather than duplicate code I made it a function.

Reply:
I think it's fine to define a function inside a function if you use it multiple times there, but nowhere else.